src/dmd/statementsem.d

```diff
@@ -4768,161 +4768,71 @@ private Statements* flatten(Statement statement, Scope* sc)
-        {
-            result = declStmt(d);
-        }
-
-        override void visit(AggregateDeclaration d)
-        {
-            result = declStmt(d);
-        }
-
-        override void visit(FuncDeclaration d)
-        {
-            result = declStmt(d);
-        }
-
-        override void visit(EnumDeclaration d)
-        {
-            result = declStmt(d);
-        }
-
-        override void visit(AliasDeclaration d)
-        {
-            result = declStmt(d);
-        }
-
-        override void visit(TemplateDeclaration d)
-        {
-            result = declStmt(d);
-        }
-
+        auto de = new DeclarationExp(s.loc, s);
+        de.type = Type.tvoid; // avoid repeated semantic
+        result = new ExpStatement(s.loc, de);
+    }
+    else if (auto d = s.isAttribDeclaration())
+    {
         /* All attributes have been already picked by the semantic analysis of
          * 'bottom' declarations (function, struct, class, etc).
          * So we don't have to copy them.
          */
-        override void visit(StorageClassDeclaration d)
+        if (Dsymbols* a = d.include(null))
         {
-            result = visitMembers(d.loc, d.decl);
-        }
-
-        override void visit(DeprecatedDeclaration d)
-        {
-            result = visitMembers(d.loc, d.decl);
-        }
-
-        override void visit(LinkDeclaration d)
-        {
-            result = visitMembers(d.loc, d.decl);
-        }
-
-        override void visit(VisibilityDeclaration d)
-        {
-            result = visitMembers(d.loc, d.decl);
-        }
-
-        override void visit(AlignDeclaration d)
-        {
-            result = visitMembers(d.loc, d.decl);
-        }
-
-        override void visit(UserAttributeDeclaration d)
-        {
-            result = visitMembers(d.loc, d.decl);
-        }
-
-        override void visit(ForwardingAttribDeclaration d)
-        {
-            result = visitMembers(d.loc, d.decl);
-        }
-
-        override void visit(StaticAssert s)
-        {
-        }
-
-        override void visit(Import s)
-        {
-        }
-
-        override void visit(PragmaDeclaration d)
-        {
-        }
-
-        override void visit(ConditionalDeclaration d)
-        {
-            result = visitMembers(d.loc, d.include(null));
-        }
-
-        override void visit(StaticForeachDeclaration d)
-        {
-            assert(d.sfe && !!d.sfe.aggrfe ^ !!d.sfe.rangefe);
-            result = visitMembers(d.loc, d.include(null));
-        }
-
-        override void visit(CompileDeclaration d)
-        {
-            result = visitMembers(d.loc, d.include(null));
+            auto statements = new Statements();
+            foreach (sx; *a)
+            {
+                statements.push(toStatement(sx));
+            }
+            result = new CompoundStatement(d.loc, statements);
         }
     }
+    else if (s.kind() == "static assert" ||  // kludge, should have isStaticAssert()
```

Comment:
`kind` returns a `const(char)*`, and `"static assert"` is a `string`.  How is this comparison even working?

Reply:
The string literal decays to a `immutable(char)*` - this check works *iff* there's exactly one instance for each string literal

Reply:
Kind of, it relies on statementsem.d and staticassert.d being compiled in the same compilation unit.  i.e: string merging of `"static assert"` means that they both point to the same address in data.

_However_ if you build the dmd front-end using separate compilation, then this will fail.